From 566289420d511f63eee242520bf23d11c97e99d2 Mon Sep 17 00:00:00 2001
From: tlhonmey <lperkins@zagmail.gonzaga.edu>
Date: Thu, 11 Oct 2018 10:25:43 -0700
Subject: [PATCH] Add support for Intel firmware RAID

---
 defaults/initrd.d/00-fsdev.sh | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/defaults/initrd.d/00-fsdev.sh b/defaults/initrd.d/00-fsdev.sh
index 29adaf49..5b9c8f88 100755
--- a/defaults/initrd.d/00-fsdev.sh
+++ b/defaults/initrd.d/00-fsdev.sh
@@ -155,6 +155,13 @@ media_find() {
 start_md_volumes() {
     good_msg "Starting md devices"
     "${MDADM_BIN}" --assemble --scan
+    # Intel Matrix RAID (and possibly others) have a container layer above the actual volumes,
+    # So we have to look for volumes that haven't been activated.
+    local DEVICE
+    for DEVICE in `cat /proc/mdstat | grep inactive | cut -d" " -f 1`; do
+      "${MDADM_BIN}" -I /dev/${DEVICE}
+    done
+
     # do not bad_msg, user could have this enabled even though
     # no RAID is currently available.
 }

