--- a/src/keyboard.c	2020-06-12 00:56:55.328680736 -0700
+++ a/src/keyboard.c	2020-06-12 00:57:43.703679101 -0700
@@ -9606,16 +9606,21 @@
 		       Fcons (make_lispy_switch_frame (frame),
 			      KVAR (interrupted_kboard, kbd_queue)));
 		  }
-                if (FIXNUMP (key) && XFIXNUM (key) != -2)
-                  {
-                    /* If interrupted while initializing terminal, we
-                       need to replay the interrupting key.  See
-                       Bug#5095 and Bug#37782.  */
-                    mock_input = 1;
-                    keybuf[0] = key;
-                  }
-                else
-                  mock_input = 0;
+		else
+		  {
+		    if (FIXNUMP (key) && XFIXNUM (key) != -2)
+		      {
+			/* If interrupted while initializing terminal, we
+			   need to replay the interrupting key.  See
+			   Bug#5095 and Bug#37782.  */
+			mock_input = 1;
+			keybuf[0] = key;
+		      }
+		    else
+		      {
+		      mock_input = 0;
+		      }
+		  }
 		goto replay_entire_sequence;
 	      }
 	  }
